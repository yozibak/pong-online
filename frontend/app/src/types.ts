export type Message = {
  username: string
  text: string
}