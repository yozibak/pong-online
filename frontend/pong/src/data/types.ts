export type PongState = {
  playerNumber: PlayerNumber
  bars: PlayerBars
  ball: Ball
  score: Score
  hasGameset: boolean
}

export type PlayerBars = Record<PlayerNumber, Bar>

export type Bar = {
  position: Position
  command: PlayerCommand
}

export type PlayerCommand = 'up' | 'down' | 'still'

export type Ball = {
  position: PongVector
  movement: PongVector
  missed?: boolean
}

export type Score = Record<PlayerNumber, number>

export type PongVector = Position & {
  set: (position: Position) => void
  angle: number
  setAngle: (angle: number) => void
  mag: number
  position: Position
}

export type Position = { x: number; y: number }

export type PlayerNumber = 1 | 2
