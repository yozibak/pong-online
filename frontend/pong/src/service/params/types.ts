export enum Params {
  GameID = 'game',
  PlayerNumber = 'player',
}