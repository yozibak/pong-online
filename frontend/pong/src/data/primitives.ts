import p5 from 'p5'
import { Ball, PongVector, Position } from './types'
import { toDegrees, toRadians } from './helpers/utils'

export const createVector = ({ x, y }: Position = { x: 0, y: 0 }): PongVector => {
  const vector = new p5.Vector(x, y)
  return {
    get x() {
      return vector.x
    },
    get y() {
      return vector.y
    },
    set(position) {
      vector.set(position.x, position.y)
    },
    
    get angle() {
      return toDegrees(vector.heading())
    },
    setAngle(angle) {
      vector.setHeading(toRadians(angle))
    },
  }
}

export const expectBallPosition = (ball: Ball) => sumPosition(ball.position, ball.movement)

export const sumPosition = (...positions: Position[]) => {
  return positions.reduce((a, b) => ({ x: a.x + b.x, y: a.y + b.y }))
}
