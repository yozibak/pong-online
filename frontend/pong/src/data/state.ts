import {
  BallVelocity,
  BarLength,
  DefaultHeight,
  DefaultWidth,
  LeftThreshold,
  RightThreshold,
} from '../config'
import { StateSnapshot } from '../domain/resolvers'
import { ReducerMap, makeStore } from './helpers/store'
import { createVector, createVectorFromAngle } from './primitives'
import { Ball, PlayMode, PlayerCommand, PlayerNumber, PongState } from './types'

export const initBall = (): Ball => ({
  position: createVector({ x: DefaultWidth / 2, y: DefaultHeight / 2 }),
  movement: createVectorFromAngle(135, BallVelocity),
})

const InitialState: PongState = {
  gameStatus: 'ready',
  playerNumber: 1,
  bars: {
    1: {
      position: {
        x: LeftThreshold,
        y: DefaultHeight / 2 - BarLength / 2,
      },
      command: 'still',
    },
    2: {
      position: {
        x: RightThreshold,
        y: DefaultHeight / 2 - BarLength / 2,
      },
      command: 'still',
    },
  },
  ball: initBall(),
  score: {
    1: 0,
    2: 0,
  },
  frameCount: 0,
  startTime: 0,
}

const reducers = {
  setPlayer: (s) => (n: PlayerNumber) => {
    s.playerNumber = n
  },
  updateCommand: (s) => (command: PlayerCommand, n?: PlayerNumber) => {
    if (n === undefined) n = s.playerNumber
    s.bars[n].command = command
  },
  setBall: (s) => (ball: Ball) => {
    s.ball = ball
  },
  incrementScore: (s) => (n: PlayerNumber) => {
    s.score[n]++
  },
  resetBall: (s) => () => {
    s.ball = initBall()
  },
  gameset: (s) => () => {
    s.gameStatus = 'gameset'
  },
  updateBySnapshot: (s) => (snapshot: StateSnapshot) => {
    s.ball = snapshot.ball
    s.bars = snapshot.bars
    s.score = snapshot.score
  },
  setStartTime: (s) => (startTime?: number) => {
    s.startTime = startTime || Date.now() + 5_000
  },
  gameStart: (s) => () => {
    s.gameStatus = 'started'
    s.ball = initBall()
  },
  incrementFrameCount: (s) => () => {
    s.frameCount++
  },
  setPlayMode: s => (mode: PlayMode) => {
    s.playMode = mode
  },
  abort: s => () => {
    s.gameStatus = 'aborted'
  }
} satisfies ReducerMap<PongState>

export const makePongStore = () => makeStore<PongState>(InitialState)(reducers)
