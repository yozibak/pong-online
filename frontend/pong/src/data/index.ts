import { makePongStore } from './state'
export * from './primitives'

export const store = makePongStore()
