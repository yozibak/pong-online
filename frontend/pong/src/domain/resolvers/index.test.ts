import { StateSnapshot, calculateNextState } from ".";
import { createVector } from "../../data";
import { Ball, Bar } from "../../data/types";

describe(`${calculateNextState.name}`, () => {
  const prepareSnapshot = ():StateSnapshot => {
    const ball: Ball = {
      position: createVector({ x: 300, y: 300,}),
      movement: createVector({ x: 10, y: 10}),
    }
    const bar1: Bar = {
      position: {
        x: 20,y: 300
      },
      command: {
        upDown: 'up'
      },
    }
    const bar2: Bar = {
      position: {
        x: 580,y: 300
      },
      command: {
        upDown: 'up'
      },
    }
    return {
      ball,
      bars: {
        1: bar1,
        2: bar2
      }
    }
  }
  it(`can calculate next state`, () => {
    expect(calculateNextState(prepareSnapshot())).toMatchSnapshot()
  })
  it(`can recursively calculate current state from old network snapshot`, () => {
    expect(calculateNextState({...prepareSnapshot(), frameAgo: 2})).toMatchSnapshot()
  })
})
