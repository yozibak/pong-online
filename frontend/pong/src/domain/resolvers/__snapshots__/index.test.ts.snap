// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculateNextState can calculate next state 1`] = `
{
  "ball": {
    "missed": undefined,
    "movement": {
      "angle": 44.99999999999999,
      "mag": 14.142135623730953,
      "position": {
        "x": 10.000000000000002,
        "y": 10,
      },
      "set": [Function],
      "setAngle": [Function],
      "x": 10.000000000000002,
      "y": 10,
    },
    "position": {
      "angle": 45,
      "mag": 438.4062043356595,
      "position": {
        "x": 310,
        "y": 310,
      },
      "set": [Function],
      "setAngle": [Function],
      "x": 310,
      "y": 310,
    },
  },
  "bars": {
    "1": {
      "command": "up",
      "position": {
        "x": 20,
        "y": 280,
      },
    },
    "2": {
      "command": "up",
      "position": {
        "x": 580,
        "y": 280,
      },
    },
  },
  "frameAgo": undefined,
}
`;

exports[`calculateNextState can recursively calculate current state from old network snapshot 1`] = `
{
  "ball": {
    "missed": undefined,
    "movement": {
      "angle": 44.99999999999999,
      "mag": 14.142135623730953,
      "position": {
        "x": 10.000000000000004,
        "y": 10,
      },
      "set": [Function],
      "setAngle": [Function],
      "x": 10.000000000000004,
      "y": 10,
    },
    "position": {
      "angle": 45,
      "mag": 452.54833995939043,
      "position": {
        "x": 320,
        "y": 320,
      },
      "set": [Function],
      "setAngle": [Function],
      "x": 320,
      "y": 320,
    },
  },
  "bars": {
    "1": {
      "command": "up",
      "position": {
        "x": 20,
        "y": 260,
      },
    },
    "2": {
      "command": "up",
      "position": {
        "x": 580,
        "y": 260,
      },
    },
  },
  "frameAgo": undefined,
}
`;
